[DROOLS-5485] Move test cases from drools-mvel to drools-test-coverage

tkobayas opened the pull request:
- First try: Copied 3 test classes

This is the very early stage of the work. I just wanted to share how I modify the test code when moving. I added you as reviewers but don't approve yet. Just feel free to share your thoughts if you have any.

**JIRA**:

https://issues.redhat.com/browse/DROOLS-5485

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

tkobayas commented on drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java:
```diff
@@ -0,0 +1,253 @@
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+public class AlphaNodeTest {
```
Basically, I will move test cases under `org.drools.mvel.integrationtests` which are usually DRL based tests. Once finished, I will likely check if there are any other good test cases to move.

tkobayas replied:
These tests are moved under `test-compiler-integration`.

I heard that Luca (DROOLS-5731) will move `drools-model-compiler` test cases to a new module under `drools-test-coverage` (not `test-compiler-integration`) so we can keep the `drools-model-compiler` test cases almost unchanged.

tkobayas commented on drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java:
```diff
@@ -0,0 +1,253 @@
+                "then\n" +
+                "end\n";
+
+        KieBase kbase = KieBaseUtil.getKieBaseFromKieModuleFromDrl("test", kieBaseTestConfiguration, str);
```
Basically, I can replace `loadKnowledgeBaseFromString(drl)` with KieBaseUtil methods. Now we can test with specified configurations (standard/flow/pattern).

lucamolteni replied:
Yes good idea, I think one of the advantages of doing this is not having many different ways to instantiate a KieSession

tkobayas commented on drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java:
```diff
@@ -0,0 +1,253 @@
+@RunWith(Parameterized.class)
+public class AlphaNodeTest {
+
+    private final KieBaseTestConfiguration kieBaseTestConfiguration;
+
+    public AlphaNodeTest(final KieBaseTestConfiguration kieBaseTestConfiguration) {
+        this.kieBaseTestConfiguration = kieBaseTestConfiguration;
+    }
+
+    @Parameterized.Parameters(name = "KieBase type={0}")
+    public static Collection<Object[]> getParameters() {
+        return TestParametersUtil.getKieBaseCloudConfigurations(true);
+    }
```
I will follow the typical approach of `test-compiler-integration`

tkobayas commented on drools-test-coverage/test-compiler-integration/src/test/java/org/drools/mvel/integrationtests/AlphaNodeTest.java:
```diff
@@ -0,0 +1,253 @@
+ * limitations under the License.
+ */
+
+package org.drools.mvel.integrationtests;
```
At the moment, I use the same package name `org.drools.mvel.integrationtests` but probably it's not suitable after finishing the move because these tests are no longer specific to mvel. `org.drools.compiler.integrationtests` is already used by existing test cases. If you have a good candidate name, let me know :)

lucamolteni replied:
I think that preserving the name might be better from the Git perspective but I'm not sure. 
While descoping I often change the package but I'm not sure it's a good idea

tkobayas replied:
Thank you for the input. Yeah, changing the name may just add extra works...

tkobayas commented:
@baldimir I added you while @hellowdan is out. Feel free to delegate to another guy if you want :)

Actually, this PR is not yet the stage of "approval". Just please share your thoughts about my approach if you have any. Thanks!

baldimir commented:
Looks good to me after a quick look.

tkobayas commented:
@baldimir  Thank you, Tibor. I'm releasing you from the reviewers as Daniel comes back :)

tkobayas commented:
Jenkins retest this

tkobayas commented:
Jenkins retest this

tkobayas commented:
Jenkins run cdb

tkobayas commented:
Jenkins retest this

tkobayas commented:
Jenkins retest this

tkobayas commented:
Hi @lucamolteni  @mariofusco  @hellowdan , Please review.

Now I moved 33 Test classes from drools-mvel to drools-test-coverage/test-compiler-integration. I arbitrarily selected these tests for this JIRA/PR as "the 1st round" because they are easy to migrate (basically, replaced CommonTestMethodBase/KieHelper methods with KieBaseUtil methods). In the future, I will evaluate remaining tests which are "worth migrating".

By this migration, now we can run these tests with executable-model (potentially, more other configurations). Even in these 33 Test classes, I found a bunch of test failures with executable-model so they are potential bugs. For this PR, I added "// TODO:" comments and run such tests without executable-model by setting `TestParametersUtil.getKieBaseCloudConfigurations(false)` in order to pass CI.

for example)
```
public class KnowledgeContextTest {
    ...

    public static Collection<Object[]> getParameters() {
     // TODO: EM failed with testKnowledgeContextMVEL. File JIRAs
        return TestParametersUtil.getKieBaseCloudConfigurations(false);
    }
```

After merging this PR, I will file JIRAs for those test failures. Once a bug is fixed (or the test failure is justified), I will remove the TODO comment and let the test run with executable-model one-by-one.

lucamolteni commented:
Thanks @tkobayas can we split the bugs found on the exec model? Let me know

lucamolteni commented on drools-test-coverage/test-compiler-integration/pom.xml:
```diff
@@ -52,6 +52,17 @@
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.drools</groupId>
```
Can we replace this will `drools-engine` and `drools-engine-classic`?

tkobayas replied:
Sure, consolidated the dependencies. I agree that it looks better for the test module.

tkobayas commented:
> Thanks @tkobayas can we split the bugs found on the exec model? Let me know

@lucamolteni Sure, I'll share the JIRAs after filing them. Thanks.

tkobayas commented:
Jenkins retest this

tkobayas commented:
> > Thanks @tkobayas can we split the bugs found on the exec model? Let me know
> 
> @lucamolteni Sure, I'll share the JIRAs after filing them. Thanks.

Filed 15 JIRAs. They are linked from https://issues.redhat.com/browse/DROOLS-5729  Feel free to take some of them if you want :)

lucamolteni commented:
@tkobayas I've taken https://issues.redhat.com/browse/DROOLS-5829

I'm starting from the bottom of the list so that we don't solve the same problem (even though they're probably related)

tkobayas commented:
> @tkobayas I've taken https://issues.redhat.com/browse/DROOLS-5829
> 
> I'm starting from the bottom of the list so that we don't solve the same problem (even though they're probably related)

@lucamolteni Thanks! That sounds good.
